nucypher/datastore/datastore.py

```diff
@@ -14,269 +14,238 @@
+        with self.__db_env.begin(write=writeable) as datastore_tx:
+            db_cursor = datastore_tx.cursor()
+
+            # Set the cursor to the closest key (if it exists) by the query params 
+            # By providing a `filter_field`, the query will immediately be
+            # limited to the subset of keys for the `filter_field`.
+            query_key = f'{record_type.__name__}:{filter_field}'.encode()
+            if not db_cursor.set_range(query_key):
+                # The cursor couldn't identify any records by the key
+                raise RecordNotFound(f"No records exist for the key from the specified query parameters: '{query_key}'")
+
+            # Check if the record at the cursor is valid for the query
+            curr_key = DatastoreKey.from_bytestring(db_cursor.key())
+            if not curr_key.compare_key(query_key):
+                raise RecordNotFound(f"No records exist for the key from the specified query parameters: '{query_key}'")
+
+            # Everything checks out, let's begin iterating!
+            # We begin by comparing the current key to the query key.
+            # If the key doesn't match the query key, we know that there are
+            # no records for the query because lmdb orders the keys lexicographically.
+            # Ergo, if the current key doesn't match the query key, we know
+            # we have gone beyond the relevant keys and can `break` the loop.
+            # Additionally, if the record is already in the `valid_records`
+            # set (identified by the `record_id`, we call `continue`.
+            for db_key in db_cursor.iternext(keys=True, values=False):
+                curr_key = DatastoreKey.from_bytestring(db_key)
+                if not curr_key.compare_key(query_key):
+                    break
+                elif curr_key.record_id in valid_records:
+                    continue
+
+                record = partial(record_type, datastore_tx, curr_key.record_id)
+                # We pass the field to the filter_func if `filter_field` and
+                # `filter_func` are both provided. In the event that the
+                # given `filter_field` doesn't exist for the record or the
+                # `filter_func` returns `False`, we call `continue`.
+                if filter_field and filter_func:
+                    try:
+                        field = getattr(record(writeable=False), filter_field)
+                    except (TypeError, AttributeError):
+                        continue
+
+                    record = partial(record_type, datastore_tx, curr_key.record_id)
+                    # We pass the field to the filter_func if `filter_field` and
+                    # `filter_func` are both provided. In the event that the
+                    # given `filter_field` doesn't exist for the record or the
+                    # `filter_func` returns `False`, we call `continue`.
+                    if filter_field and filter_func:
```

Comment:
isn't this check redundant based on L199?

Reply:
Oh good catch! The check isn't just redundant, the whole block is!

I must've not deleted this block during the rebase.

Reply:
Resolved in 078b5b1